### DISCUSSION TITLE
custom server with custom routing

### DISCUSSION THREAD

**Discussion by alicerocheman** [vercel/next.js#23579] [Help]
TLDR; using custom server and app.render, my pageProps are empty, and a href/as mismatch error is thrown (I'm not using Link)

Hello everyone 

I'm building a custom server in my next app, for several reasons, one of them being that we don't want our urls to exactly match our pages folders.

pages file architecture:
```
pages /
| entity /
| | entityType1 /
| | | - [...slug].tsx
| | entityType2 /
| | | - [...slug]
...
```
desired urls : 
`www.mydomain.com/entityType1/id/name`
`www.mydomain.com/entityType2/id/name`

In other words, we don't want "entity" appearing in the url, but we need the folder in pages, because entity carries common code for the varied entityTypes...

So I created the following custom server:
```ts
import { createServer } from 'http';
import { parse } from 'url';
import next from 'next';
import getConfig from './getConfig';

const port = parseInt(process.env.PORT || '3000', 10);
const dev = process.env.NODE_ENV !== 'production';
const conf = getConfig(dev);
const app = next({ dev, conf });
const handle = app.getRequestHandler();

const ENTITY_PAGES = [
  'entityType1',
  'entityType2',
  'entityType3',
  'entityType4',
];

app.prepare().then(() => {
  createServer((req, res) => {
    /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */
    const parsedUrl = parse(req.url!, true);
    const { pathname, query } = parsedUrl;

    const pathelements = pathname?.split('/');

    if (pathelements && pathelements.length > 2) {
      if (ENTITY_PAGES.includes(pathelements[1])) {
        pathelements.splice(1, 0, 'entity');
        const newPathname = pathelements.join('/');
        app.render(req, res, newPathname, query);
      } else {
        handle(req, res, parsedUrl);
      }
    } else {
      handle(req, res, parsedUrl);
    }
  }).listen(port, () => {
    console.log(
      `> Ready on http://localhost:${port} as ${dev ? 'development' : process.env.NODE_ENV}`
    );
  });
});
```

so the query `/entityTypeX/id/name` should render the page `/entity/entityTypeX/id/name`.

My issue is I get an empty pageProps (=== {}) when calling `/entityTypeX/id/name`

The error in the browser console says that 
```
The provided `as` value (/entityType1/15/ch%C3%A2teau-alice) is incompatible with the `href` value (/entity/entityType1/[...id_name])
```
and points to a Next Link documentation, but I'm not using Link, I'm typing the url myself in the browser...

Anyway I'm at a loss, and would welcome any pointers you may give me!
